File: Greedy/Java/Divide Intervals Into Minimum Number of Groups.java
// Problem link: https://leetcode.com/problems/divide-intervals-into-minimum-number-of-groups/description/
// Time: O(n + k)
// Space: O(k)
class Solution {
    public int minGroups(int[][] intervals) {
        int minStart = Integer.MAX_VALUE, maxEnd = Integer.MIN_VALUE, result = 0, prefix = 0;

        for(int[] interval : intervals) {
            minStart = Math.min(interval[0], minStart);
            maxEnd = Math.max(interval[1], maxEnd);
        }

        int[] lineSweep = new int[maxEnd + 1];

        for(int[] interval : intervals) {
            lineSweep[interval[0] - 1]++;
            lineSweep[interval[1]]--;
        }

        for(int point = minStart - 1; point <= maxEnd; point++) {
            prefix += lineSweep[point];
            result = Math.max(prefix, result);
        }

        return result;
    }
}

File: Greedy/Python/Divide Intervals Into Minimum Number of Groups.py
# Problem link: https://leetcode.com/problems/divide-intervals-into-minimum-number-of-groups/description/
# Time: O(n + k)
# Space: O(k)
class Solution:
    def minGroups(self, intervals: List[List[int]]) -> int:
        minStart = float('inf')
        maxEnd = float('-inf')

        for start, end in intervals:
            minStart = min(minStart, start)
            maxEnd = max(maxEnd, end)
        
        lineSweep = [0] * (maxEnd + 1)

        for start, end in intervals:
            lineSweep[start - 1] += 1
            lineSweep[end] -= 1
        
        result = prefix = 0

        for point in range(minStart - 1, maxEnd + 1):
            prefix += lineSweep[point]
            result = max(result, prefix)
        
        return result
